bhttp/src/http_compat/decode.rs

```diff
@@ -0,0 +1,538 @@
+    /// and constructs the corresponding HTTP request or response parts.
+    ///
+    /// # Arguments
+    ///
+    /// * `header` - The BHTTP header to convert
+    ///
+    /// # Returns
+    ///
+    /// Either HTTP request parts or response parts, depending on the control data,
+    /// or an error if parsing fails.
+    fn try_from(header: crate::Header) -> Res<Self> {
+        if header.control().is_request() {
+            Ok(HttpMessageParts::Request(request::Parts::try_from(header)?))
+        } else {
+            Ok(HttpMessageParts::Response(response::Parts::try_from(
+                header,
+            )?))
+        }
+    }
+}
+
+impl TryFrom<crate::Header> for request::Parts {
+    type Error = Error;
+
+    fn try_from(header: crate::Header) -> Res<Self> {
+        let control_data = header.control();
+
+        // Build the HTTP request parts
+        let mut builder = request::Builder::new();
+
+        // Set method
+        if let Some(method) = control_data.method() {
+            builder = builder.method(method);
+        }
+
+        // Convert field section to headers
+        let headers = HeaderMap::try_from(&header.fields)?;
+
+        // Set URI
+        let uri = {
+            let mut uri_builder = Uri::builder();
+
+            if let Some(scheme) = control_data.scheme() {
+                if !scheme.is_empty() {
+                    uri_builder = uri_builder.scheme(scheme);
+                }
+            }
+
+            // Handle authority field
+            match control_data.authority() {
+                Some(authority_bytes) if !authority_bytes.is_empty() => {
+                    uri_builder = uri_builder.authority(std::str::from_utf8(authority_bytes)?);
+                }
+                _ => {
+                    // The http::Uri does not allow an absent authority when the scheme is present, but
+                    // this is permitted in bhttp and report an InvalidUri(InvalidFormat) error. To
+                    // reconcile this difference, we will attempt to retrieve the authority from the
+                    // request headers.
+                    if let Some(scheme) = control_data.scheme() {
+                        if !scheme.is_empty() {
+                            if let Some(host_value) = headers.get("host") {
+                                uri_builder = uri_builder.authority(host_value.to_str()?);
+                            }
+                        }
+                    }
+                }
+            }
+
+            if let Some(path) = control_data.path() {
+                if !path.is_empty() {
+                    uri_builder = uri_builder.path_and_query(std::str::from_utf8(path)?);
+                }
+            }
+
+            uri_builder.build()?
+        };
+        builder = builder.uri(uri);
+
+        // Set headers
+        for (name, value) in headers {
+            if let Some(name) = name {
+                builder = builder.header(name, value);
+            }
```

Comment:
You might want to filter out pseudo-headers here.  Most of these are in the control data, but the format allows for others.  I don't know if the http crate is OK with field names that start with ':', but it looks like they will produce an error at some level, quite possibly a panic.